File: Merge Two Sorted Lists/merge_sorted_lists.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	fmt.Println("Hello world")
	l1 := createListNode([]int{-6, -5, 6, 6, 7})
	l2 := createListNode([]int{0})
	l3 := mergeTwoLists(&l1, &l2)
	fmt.Println(stringList(l3))
}

func stringList(node *ListNode) string {
	var listString strings.Builder
	for node != nil {
		val := strconv.Itoa(node.Val)
		listString.WriteString(val)
		if node.Next != nil {
			listString.WriteString(" -> ")
		}
		node = node.Next
	}
	return listString.String()
}

func createListNode(list []int) ListNode {
	node := &ListNode{list[0], nil}
	head := node
	for _, num := range list[1:] {
		node.Next = &ListNode{num, nil}
		node = node.Next
	}
	return *head
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	current := dummy
	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			current.Next = list1
			list1 = list1.Next
		} else {
			current.Next = list2
			list2 = list2.Next
		}
		current = current.Next
	}

	if list1 != nil {
		current.Next = list1
	} else if list2 != nil {
		current.Next = list2
	}

	return dummy.Next
}

File: Merge Two Sorted Lists/merge_sorted_lists.exs
defmodule ListNode do
  @type t :: %__MODULE__{
          val: integer,
          next: ListNode.t() | nil
        }
  defstruct val: 0, next: nil
end

defmodule Solution do
  @spec merge_two_lists(list1 :: ListNode.t() | nil, list2 :: ListNode.t() | nil) ::
          ListNode.t() | nil
  def merge_two_lists(list1, list2) do
    head = %ListNode{}
    merge_two_lists(head, list1, list2)
    head.next
  end

  def merge_two_lists(head, nil, nil), do: nil
  def merge_two_lists(head, list1, nil), do: head = %{head | next: list1}
  def merge_two_lists(head, nil, list2), do: head = %{head | next: list2}

  def merge_two_lists(head, list1, list2) do
    if list1.val <= list2.val do
      head = %{head | val: list1.val}
      head = %{head | next: merge_two_lists(%ListNode{}, )}
    else
      2
    end
  end

  def create_list_node([head | []]) do
    %ListNode{val: head}
  end

  def create_list_node([head | tail]) do
    node = %ListNode{val: head, next: Solution.create_list_node(tail)}
  end
end

l1 = Solution.create_list_node([1, 2, 3])
IO.inspect(l1)
l2 = Solution.create_list_node([1, 3, 4])
IO.inspect(l2)

Solution.merge_two_lists(l1, l2)
